orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java

```diff
@@ -0,0 +1,231 @@
+    static final OrcValReader<?> INSTANCE = new BooleanReader();
+
+    private BooleanReader() {
+    }
+
+    @Override
+    public Boolean nonNullRead(ColumnVector vector, int row) {
+      return ((LongColumnVector) vector).vector[row] != 0;
+    }
+  }
+
+  private static class ShortReader implements OrcValReader<Short> {
+    static final OrcValReader<?> INSTANCE = new ShortReader();
+
+    private ShortReader() {
+    }
+
+    @Override
+    public Short nonNullRead(ColumnVector vector, int row) {
+      return (short) ((LongColumnVector) vector).vector[row];
+    }
+  }
+
+  private static class IntegerReader implements OrcValReader<Integer> {
+    static final OrcValReader<?> INSTANCE = new IntegerReader();
+
+    private IntegerReader() {
+    }
+
+    @Override
+    public Integer nonNullRead(ColumnVector vector, int row) {
+      return (int) ((LongColumnVector) vector).vector[row];
+    }
+  }
+
+  private static class LongReader implements OrcValReader<Long> {
+    static final OrcValReader<?> INSTANCE = new LongReader();
+
+    private LongReader() {
+    }
+
+    @Override
+    public Long nonNullRead(ColumnVector vector, int row) {
+      return ((LongColumnVector) vector).vector[row];
+    }
+  }
+
+  private static class FloatReader implements OrcValReader<Float> {
+    private static final FloatReader INSTANCE = new FloatReader();
+
+    private FloatReader() {
+    }
+
+    @Override
+    public Float nonNullRead(ColumnVector vector, int row) {
+      return (float) ((DoubleColumnVector) vector).vector[row];
+    }
+  }
+
+  private static class DoubleReader implements OrcValReader<Double> {
+    private static final DoubleReader INSTANCE = new DoubleReader();
+
+    private DoubleReader() {
+    }
+
+    @Override
+    public Double nonNullRead(ColumnVector vector, int row) {
+      return ((DoubleColumnVector) vector).vector[row];
+    }
+  }
+
+  private static class ByteReader implements OrcValReader<Byte> {
+    private static final ByteReader INSTANCE = new ByteReader();
+
+    private ByteReader() {
+    }
+
+    @Override
+    public Byte nonNullRead(ColumnVector vector, int row) {
+      return (byte) ((LongColumnVector) vector).vector[row];
+    }
+  }
+
+  private static class BytesReader implements OrcValReader<byte[]> {
+    private static final BytesReader INSTANCE = new BytesReader();
+
+    private BytesReader() {
+    }
+
+    @Override
+    public byte[] nonNullRead(ColumnVector vector, int row) {
+      BytesColumnVector bytesVector = (BytesColumnVector) vector;
+
+      return Arrays.copyOfRange(
+          bytesVector.vector[row], bytesVector.start[row], bytesVector.start[row] + bytesVector.length[row]);
+    }
+  }
+
+  public abstract static class StructReader<T> implements OrcValReader<T> {
+    private final OrcValReader<?>[] readers;
+    private final int[] positions;
+    private final Object[] constants;
+
+    protected StructReader(List<OrcValReader<?>> readers) {
```

Comment:
Where is this used? It doesn't look like using 0-length arrays for positions and constants is correct.

Reply:
Nevermind, I see what's happening.